### PR TITLE
Refactor fetcher config

### PR DESCRIPTION
**Pull request by ajbouh**
DRY up fetcher configuration

Consolidate the addition of the fetcher container to the pod into a new type: FetcherConfig.

Add FissionFetcherSA as pod serviceAccountName if none is set

Add PreStop lifecycle handler if pod set TerminationGracePeriodSeconds

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/1150)

### CODE REVIEW

File: environments/fetcher/config/config.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+func (cfg *Config) SharedMountPath() string {
+	return cfg.sharedMountPath
+}
+
+func (cfg *Config) NewSpecializeRequest(fn *crd.Function, env *crd.Environment) fission.FunctionSpecializeRequest {
+	// for backward compatibility, since most v1 env
+	// still try to load user function from hard coded
+	// path /userfunc/user
+	targetFilename := "user"
+	if env.Spec.Version == 2 {
```
[the remark this answers is not in the document]
Comment (ajbouh) [reply environments/fetcher/config/config.go]: Will do.

**Comment by ajbouh** [comment]
Aside from addressing the comment above, what's the next step on this?

File: environments/fetcher/config/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ajbouh) [reply environments/fetcher/config/config.go]:
```suggestion
	if env.Spec.Version >= 2 {
```

**Comment by soamvasani** [comment]
> Aside from addressing the comment above, what's the next step on this?

Needs a rebase to latest master.

I started a travis ci run: https://travis-ci.org/fission/fission/builds/521434792. If no bugs are found there, we just merge after the rebase.

**Comment by soamvasani** [comment]
build failed with:
```
!!! 'gofmt -s' needs to be run on the following files: 
./environments/fetcher/config/config.go
```

File: executor/poolmgr/gp.go [hunk 2]
```diff
@@ -92,6 +86,7 @@ func MakeGenericPool(
 	namespace string,
 	functionNamespace string,
 	fsCache *fscache.FunctionServiceCache,
+	fetcherConfig *fetcherConfig.Config,
 	instanceId string,
 	enableIstio bool,
 	collectorEndpoint string) (*GenericPool, error) {
```
Comment (soamvasani) [inline executor/poolmgr/gp.go]: Did you mean to remove `collectorEndpoint` from here?

File: buildermgr/envwatcher.go [hunk 3]
```diff
@@ -101,12 +105,6 @@ func makeEnvironmentWatcher(logger *zap.Logger, fissionClient *crd.FissionClient
...
 	}
 
-	fetcherImage := os.Getenv("FETCHER_IMAGE")
-	if len(fetcherImage) == 0 {
-		fetcherImage = "fission/fetcher"
-	}
-
-	fetcherImagePullPolicy := fission.GetImagePullPolicy(os.Getenv("FETCHER_IMAGE_PULL_POLICY"))
 	builderImagePullPolicy := fission.GetImagePullPolicy(os.Getenv("BUILDER_IMAGE_PULL_POLICY"))
 	collectorEndpoint := os.Getenv("TRACE_JAEGER_COLLECTOR_ENDPOINT")
```
Comment (soamvasani) [inline buildermgr/envwatcher.go]: unused, causes build to fail

```suggestion
```

File: executor/poolmgr/gp.go (hunk 2 above)
Comment (ajbouh) [reply executor/poolmgr/gp.go]: Yes, good catch

**Comment by vishal-biyani** [comment]
Closed by https://github.com/fission/fission/pull/1168, thanks a lot @ajbouh for this work!

**Pull request closed without merging**